062-topic-replication-factor-change.md

```diff
@@ -0,0 +1,223 @@
+
+At runtime, the Topic Operator watches and periodically checks all managed and non-paused `KafkaTopic` resources.
+Among the other configurations, the reconciliation logic also search for `.spec.replicas` update by comparing with the actual topic replication factor.
+From the point of view of the Topic Operator, a topic replication factor change can be in the following states:
+
+- Pending: `.spec.replicas` is updated, but the configuration change request is not sent to Cruise Control (`generation` > `observedGeneration`).
+- Ongoing: `.spec.replicas` is updated and Cruise Control has accepted the request, but the task is not completed (`generation` == `observedGeneration`).
+- Failed: `.spec.replicas` is updated with an invalid value, or there is a hard goal violation, or Cruise Control is not available (`generation` == `observedGeneration`).
+
+To request a replication factor change, the Topic Operator needs to authenticate and send HTTP requests to the Cruise Control's REST API.
+For that reason, we introduce the following environment variables, which will be set automatically by the Cluster Operator, or manually by the user in case of standalone deployment.
+
+| Name                                   | Type    | Default | Description                                           |
+|:---------------------------------------|---------|---------|-------------------------------------------------------|
+| STRIMZI_CRUISE_CONTROL_ENABLED         | Boolean | false   | Whether Cruise Control integration is enabled         |
+| STRIMZI_CRUISE_CONTROL_NAMESPACE       | String  | ""      | Namespace where Cruise Control is running             |
+| STRIMZI_CRUISE_CONTROL_HOSTNAME        | String  | ""      | Cruise Control hostname                               |
+| STRIMZI_CRUISE_CONTROL_PORT            | Integer | 9090    | Cruise Control port                                   |
+| STRIMZI_CRUISE_CONTROL_SSL_ENABLED     | Boolean | false   | Whether Cruise Control SSL encryption is enabled      |
+| STRIMZI_CRUISE_CONTROL_AUTH_ENABLED    | Boolean | false   | Whether Cruise Control Basic authorization is enabled |
+| STRIMZI_CRUISE_CONTROL_SECRET_NAME     | String  | ""      | Secret name containing Cruise Control truststore      |
+| STRIMZI_CRUISE_CONTROL_API_SECRET_NAME | String  | ""      | Secret name containing Cruise Control API credentials |
+| STRIMZI_CRUISE_CONTROL_RACK_ENABLED    | Boolean | false   | Whether the target Kafka cluster has rack awareness   |
+
+The topic configuration endpoint is asynchronous, which means that it validates the task, but the execution will be queued.
+Inside Cruise Control, a valid topic configuration task is first create in an `Active` state, then moved to `InExecution`, and finally to `Completed` or `CompletedWithError`.
+Only a single task can be `InExecution` state at any given time, which includes both topic configuration and rebalance tasks.
```

Comment:
How will users monitor what is going on? It might be quite confusing when they try to scale-down or do a rebalance and it is blocked because some RF changes queued by the TO. So it is important to make it easy to understand what is happening. Will CO know about these tasks and be able to log something like `CC is busy with another rebalance`? How will the queue be _cleaned_ when CC gets stuck for some reason?

Reply:
The mechanism is always the same, I use `Warning` status and logs to inform the user about the current status of the replication factor change. CC uses an in-memory task queue, so if it gets stuck, restarting CC should unblock, but it never happened to me.

Reply:
Yeah ... but imagine ... the KafkaRebalance is stuck, there are several KafkaTopic changing replication factor, etc. How does the user find out what is going on? Maybe it has no easy solution, but it would be good to understand the workflow.

Reply:
CC has a number of configurable timeouts (i.e. `admin.client.request.timeout.ms`, `leader.movement.timeout.ms`) and also alert thresholds. Any stuck task will trigger an error on timeout that will be reflected in CC logs and specific resource status. After that, any queued task can be executed, one at a time.